Makefile

```diff
@@ -17,144 +17,170 @@
+# Initial output
+$(info $(shell ${ECHOE} ${GREEN}TARGET_EXEC [$(TARGET_EXEC)]${NC}))
+$(info TARGET [$(TARGET)])
+$(info COMPILER [$(COMPILER)])
+$(info $(shell ${ECHOE} ${GREEN}VARIANT [$(VARIANT)]${NC}))
 
+# Search for prerequisites (like .c, .cpp, or .h files) in specified directories when those files are not in the current directory.
+VPATH += $(UT_CORE_DIR) $(TOP_DIR)
+
+# Default target
 .PHONY: clean list arm linux framework test createdirs all printenv
 
 all: framework $(OBJS)
 
-# Ensure the framework is built
+# Build framework
 # Recursive make is needed as src files are not available during the first iteration
-framework: createdirs download_and_build
+framework: checkvariantchange createdirs download_and_build
 	@${ECHOE} ${GREEN}Framework downloaded and built${NC}
-	@make test
-	@cp $(UT_CONTROL)/build/$(TARGET)/lib/libut_control.* ${LIB_DIR}
-	@cp $(UT_CONTROL)/build/$(TARGET)/lib/libut_control.* ${BIN_DIR}
+	@make test VARIANT=${VARIANT}
+	@cp -r $(UT_CONTROL)/build/$(TARGET)/lib/libut_control.* $(LIB_DIR) $(BIN_DIR)
 	@${ECHOE} ${GREEN}ut-control LIB Copied to [${BIN_DIR}]${NC}
 
 download_and_build:
-	@${ECHOE} ${GREEN}"Ensure ut-core frameworks are present"${NC}
-	@${UT_CORE_DIR}/build.sh TARGET=${TARGET}
+	@${ECHOE} ${GREEN}Ensure ut-core frameworks are present${NC}
+	@${UT_CORE_DIR}/build.sh TARGET=${TARGET} VARIANT=${VARIANT}
 	@${ECHOE} ${GREEN}Completed${NC}
-	@${ECHOE} ${GREEN}"Entering ut-control [TARGET=${TARGET}]"${NC}
+	@${ECHOE} ${GREEN}Entering ut-control [TARGET=${TARGET}]${NC}
 	@${MAKE} -C $(UT_CONTROL) TARGET=${TARGET}
 
-# Make the final test binary
+# Build the test binary
 test: $(OBJS) createdirs
 	@${ECHOE} ${GREEN}Linking $@ $(BUILD_DIR)/$(TARGET_EXEC)${NC}
-	@$(CC) $(OBJS) -o $(BUILD_DIR)/$(TARGET_EXEC) $(XLDFLAGS) $(KCFLAGS) $(XCFLAGS)
+	@$(COMPILER) $(OBJS) -o $(BUILD_DIR)/$(TARGET_EXEC) $(XCFLAGS) $(XLDFLAGS)
 	@cp $(BUILD_DIR)/$(TARGET_EXEC) $(BIN_DIR)/
 ifneq ("$(wildcard $(HAL_LIB_DIR)/*.so)","")
 	cp $(HAL_LIB_DIR)/*.so* $(BIN_DIR)
 endif
 
+# Create necessary directories
 createdirs:
-	@$(MKDIR_P) ${BIN_DIR}
-	@$(MKDIR_P) ${LIB_DIR}
+	@echo "$(VARIANT)" > $(VARIANT_FILE)
+	@$(MKDIR_P) ${BIN_DIR} ${LIB_DIR}
 
-# Make any c source
+# Compilation rules
 $(BUILD_DIR)/%.o: %.c
 	@${ECHOE} ${GREEN}Building [${YELLOW}$<${GREEN}]${NC}
 	@$(MKDIR_P) $(dir $@)
-	@$(CC) $(XCFLAGS) -c $< -o $@
+	@$(COMPILER) $(XCFLAGS) -c $< -o $@
+
+$(BUILD_DIR)/%.o: %.cpp
+	@${ECHOE} ${GREEN}Building [${YELLOW}$<${GREEN}]${NC}
+	@$(MKDIR_P) $(dir $@)
+	@$(COMPILER) $(CXXFLAGS) -c $< -o $@
+
+# Rule to check if the VARIANT has changed
+checkvariantchange:
+	@if [ -f "$(VARIANT_FILE)" ]; then \
+		PREVIOUS_VARIANT=$$(cat $(VARIANT_FILE)); \
+		if [ "$$PREVIOUS_VARIANT" != "$(VARIANT)" ]; then \
+			${ECHOE} ${RED}********************************************${NC}; \
+			${ECHOE} ${RED}Error: VARIANT has changed from $$PREVIOUS_VARIANT to $(VARIANT)${NC}; \
+			${ECHOE} ${RED}Please clean the build and try again.${NC}; \
```

Comment:
you should be specific about what you mean..

please run `make clean TARGET=${TARGET}`

Is there anything else required?

Reply:
done